pages/api/webhooks/stripe.ts

```diff
@@ -40,41 +47,102 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+          'missing Email from Stripe webhook',
+          event,
+        );
+        throw new Error('missing Email from Stripe webhook');
+      }
 
-        // Only store expiration for one month plan
+      const sinceDate = dayjs.unix(session.created).utc().toDate();
+      const proPlanExpirationDate = await (async () => {
+        // Takes plan_giving_weeks priority over plan_code
+        if (planGivingWeeks && typeof planGivingWeeks === 'string') {
+          // Get users' pro expiration date
+          const supabase = getAdminSupabaseClient();
+          const user = await userProfileQuery({
+            client: supabase,
+            email,
+          });
+          const userProPlanExpirationDate = user?.proPlanExpirationDate;
+          if (userProPlanExpirationDate) {
+            // when user bought one-time pro plan previously or user has referral trial
+            return dayjs(userProPlanExpirationDate)
+              .add(+planGivingWeeks, 'week')
+              .toDate();
+          } else if (user.plan === 'pro' && !user.proPlanExpirationDate) {
+            // when user is pro monthly subscriber
+            await sendReportForStripeWebhookError(
+              'Monthly Pro subscriber bought one-time pro plan',
+              event,
+              user,
+            );
+
+            throw new Error(
+              'Monthly Pro subscriber bought one-time pro plan, should not happen',
+            );
+          } else {
+            // when user is not pro yet
+            return dayjs(sinceDate).add(+planGivingWeeks, 'week').toDate();
+          }
+        } else if (planCode === ONE_TIME_PRO_PLAN_FOR_1_MONTH) {
+          // Only store expiration for one month plan
+          return dayjs(sinceDate).add(1, 'month').toDate();
+        } else {
+          return undefined;
+        }
+      })();
+
+      // Update user account by User id
+      if (userId) {
         await updateUserAccount({
           upgrade: true,
           userId,
           stripeSubscriptionId,
-          proPlanExpirationDate: planCode === ONE_TIME_PRO_PLAN_FOR_1_MONTH ? proPlanExpirationDateUTC : undefined,
+          proPlanExpirationDate: proPlanExpirationDate,
+        });
+      } else {
+        // Update user account by Email
+        await updateUserAccount({
+          upgrade: true,
+          email: email!,
+          stripeSubscriptionId,
+          proPlanExpirationDate: proPlanExpirationDate,
         });
       }
     }
 
+    // Monthly Pro Plan Subscription recurring payment
     if (event.type === 'customer.subscription.updated') {
       const session = event.data.object as Stripe.Subscription;
       stripeSubscriptionId = session.id;
 
       if (!stripeSubscriptionId) {
-        console.error(session);
+        await sendReportForStripeWebhookError(
+          'Subscription id not found from Stripe webhook',
+          event,
+        );
         throw new Error('Subscription id not found from Stripe webhook');
```

Comment:
Can we also use customer email to look for them, if the stripe subscriptionID is not present?